File: sdks/go/loreweave_mcp/meta.go
package loreweave_mcp

import (
	"fmt"

	"github.com/modelcontextprotocol/go-sdk/mcp"
)

// C-TOOL `_meta` convention. Every kit-registered tool MUST carry machine-readable
// metadata so the consumer (auto-apply vs confirm), the gateway (prefix), and the
// FE behave correctly without hardcoding tool names. Enforcement is NET-NEW in the
// kit: a tool registered without a valid tier AND scope is rejected (legacy
// glossary/knowledge tools predate `_meta` and are exempt — only kit-registered
// providers go through ValidateToolMeta / MustValidateToolMeta).
//
// The metadata lives in mcp.Tool.Meta (the `_meta` JSON object), under these keys:
const (
	MetaKeyTier       = "tier"       // R|A|W|S — drives auto-apply vs confirm
	MetaKeyScope      = "scope"      // book|project|user|none — drives which guard runs
	MetaKeyUndoHint   = "undo_hint"  // optional {tool, args} for C-ACTIVITY
	MetaKeySynonyms   = "synonyms"   // optional alias terms feeding find_tools (H6)
	MetaKeyVisibility = "visibility" // discoverable|legacy — CAT-4 (mcp-tool-io.md Part 4)
	MetaKeyAsync      = "async"      // true ⇒ tool STARTS a background job (async-honesty)
	MetaKeyPaid       = "paid"       // true ⇒ calling it SPENDS real money (Track D CD1)
	// MetaKeySupersededBy — the tool that replaces a `legacy` one. Consumers already
	// read it (`tool_list`/`tool_load` label it); until now NOTHING produced it.
	MetaKeySupersededBy = "superseded_by"
)

// Visibility is the CAT-4 catalog-hygiene enum. Absent (zero value) reads as
// VisibilityDiscoverable — a tool with no explicit visibility is discoverable by
// default, so existing tools need no change. A "legacy" tool keeps its schema and
// handler working for any existing caller; it is excluded from find_tools/
// search_catalog and from any domain hot-seed on BOTH federation surfaces
// (chat-service tool_discovery.py + ai-gateway find-tools.ts must stay in
// lockstep — see mcp-tool-io.md CAT-4). The ONLY path back to a legacy tool is
// an explicit, user-initiated per-session pin (Settings & Configuration Boundary
// SET-1) — never a blanket unlock.
type Visibility string

const (
	VisibilityDiscoverable Visibility = "discoverable"
	VisibilityLegacy       Visibility = "legacy"
)

// WithVisibility returns a copy of m with _meta.visibility set. A plain function,
// not a method — mcp.Meta is a map type owned by the go-sdk package, so we can't
// attach methods to it from here. Callers chain it onto NewToolMeta's result:
//
//	Meta: lwmcp.WithVisibility(lwmcp.NewToolMeta(lwmcp.TierA, lwmcp.ScopeBook, nil, nil), lwmcp.VisibilityLegacy)
func WithVisibility(m mcp.Meta, v Visibility) mcp.Meta {
	m[MetaKeyVisibility] = string(v)
	return m
}

// WithAsync returns a copy of m with _meta.async=true — the tool STARTS a background
// job (queued; not done when the call returns). Chained onto NewToolMeta like
// WithVisibility. This is the DURABLE async-honesty signal: a consumer (the workflow
// step-runner) reads it from the catalog instead of guessing from the tool name.
//
//	Meta: lwmcp.WithAsync(lwmcp.NewToolMeta(lwmcp.TierW, lwmcp.ScopeBook, nil, nil))
func WithAsync(m mcp.Meta) mcp.Meta {
	m[MetaKeyAsync] = true
	return m
}

// WithPaid returns a copy of m with _meta.paid=true — calling this tool SPENDS REAL
// MONEY (an outward paid API, or LLM/embedding tokens). Track D CD1.
//
// `paid` is ORTHOGONAL to `tier`: spend governs money, tier governs mutation. A PAID
// READ (e.g. web search) is legitimate — it stays tier R and remains callable in `ask`
// mode — but it must clear a SPEND gate, never a write gate. Do NOT coerce a tool to
// tier A/W merely because it costs money.
//
//	Meta: lwmcp.WithPaid(lwmcp.NewToolMeta(lwmcp.TierR, lwmcp.ScopeNone, nil, nil))
func WithPaid(m mcp.Meta) mcp.Meta {
	m[MetaKeyPaid] = true
	return m
}

// WithSupersededBy returns a copy of m with _meta.superseded_by set — names the tool
// that REPLACES this one. Pair it with WithVisibility(..., VisibilityLegacy) when a tool
// is renamed: the old name keeps working (deprecate, never delete — CAT-4) and both
// `tool_list` and `tool_load` label it with its replacement so an agent migrates itself.
//
//	Meta: lwmcp.WithSupersededBy(
//	        lwmcp.WithVisibility(lwmcp.NewToolMeta(...), lwmcp.VisibilityLegacy),
//	        "web_search")
func WithSupersededBy(m mcp.Meta, tool string) mcp.Meta {
	m[MetaKeySupersededBy] = tool
	return m
}

// Tier is the C-TOOL tier enum: R(ead) A(uto) W(rite-confirm) S(chema/secret).
type Tier string

const (
	TierR Tier = "R" // read — no confirm, doesn't count against the write budget
	TierA Tier = "A" // auto-commit + activity/Undo strip
	TierW Tier = "W" // confirm_action (write)
	TierS Tier = "S" // confirm_action (schema/secret)
)

// Scope is the C-TOOL scope enum, selecting which kit guard a tool's handler runs.
type Scope string

const (
	ScopeBook    Scope = "book"
	ScopeProject Scope = "project"
	ScopeUser    Scope = "user"
	ScopeNone    Scope = "none"
)

func validTier(s string) bool {
	switch Tier(s) {
	case TierR, TierA, TierW, TierS:
		return true
	default:
		return false
	}
}

func validScope(s string) bool {
	switch Scope(s) {
	case ScopeBook, ScopeProject, ScopeUser, ScopeNone:
		return true
	default:
		return false
	}
}

// ValidateToolMeta rejects a tool whose `_meta` is missing or carries an invalid
// tier or scope. It is the C-TOOL gate every kit-registered provider runs before
// AddTool. Returns a descriptive error naming the offending tool + field.
func ValidateToolMeta(t *mcp.Tool) error {
	if t == nil {
		return fmt.Errorf("loreweave_mcp: nil tool")
	}
	if t.Meta == nil {
		return fmt.Errorf("loreweave_mcp: tool %q has no _meta (tier+scope required)", t.Name)
	}
	tierVal, ok := t.Meta[MetaKeyTier].(string)
	if !ok || !validTier(tierVal) {
		return fmt.Errorf("loreweave_mcp: tool %q _meta.tier missing or invalid (want R|A|W|S, got %v)", t.Name, t.Meta[MetaKeyTier])
	}
	scopeVal, ok := t.Meta[MetaKeyScope].(string)
	if !ok || !validScope(scopeVal) {
		return fmt.Errorf("loreweave_mcp: tool %q _meta.scope missing or invalid (want book|project|user|none, got %v)", t.Name, t.Meta[MetaKeyScope])
	}
	return nil
}

// MustValidateToolMeta panics on an invalid `_meta`. Use it at registration time
// (server construction) where a malformed tool is a programming error that should
// fail the build/boot, not a runtime condition.
func MustValidateToolMeta(t *mcp.Tool) {
	if err := ValidateToolMeta(t); err != nil {
		panic(err)
	}
}

// NewToolMeta builds a C-TOOL-compliant mcp.Meta with the required tier+scope and
// optional extras, so providers don't hand-assemble the map (and can't typo a
// key). Pass nil/"" for unused optionals.
func NewToolMeta(tier Tier, scope Scope, undoHint map[string]any, synonyms []string) mcp.Meta {
	m := mcp.Meta{
		MetaKeyTier:  string(tier),
		MetaKeyScope: string(scope),
	}
	if undoHint != nil {
		m[MetaKeyUndoHint] = undoHint
	}
	if len(synonyms) > 0 {
		m[MetaKeySynonyms] = synonyms
	}
	return m
}

File: sdks/python/loreweave_mcp/meta.py
"""Per-tool `_meta` validator (C-TOOL enforcement) — built fresh.

Every kit-registered tool MUST carry machine-readable metadata so the consumer,
gateway, and FE behave correctly without hardcoding tool names:

  - ``_meta.tier``  ∈ {R, A, W, S}   — drives auto-apply vs. confirm.
  - ``_meta.scope`` ∈ {book, project, user, none} — drives which guard runs.
  - ``_meta.undo_hint`` (optional)   — {tool, args} for the Tier-A activity strip.
  - ``_meta.synonyms`` (optional)    — alias terms feeding find_tools recall.

The kit REJECTS a tool registered without BOTH ``tier`` and ``scope`` (legacy
glossary/knowledge tools predate `_meta` and are exempt; only kit-registered
providers must carry it). `require_meta` builds the validated `_meta` dict to pass
to `@server.tool(..., meta=require_meta(...))`.
"""

from __future__ import annotations

from typing import Any

__all__ = [
    "TIERS",
    "SCOPES",
    "MetaValidationError",
    "validate_tool_meta",
    "require_meta",
]

TIERS = frozenset({"R", "A", "W", "S"})
SCOPES = frozenset({"book", "project", "user", "none"})


class MetaValidationError(ValueError):
    """A tool was registered with missing/invalid ``_meta`` (no tier or scope, or
    a value outside the allowed enum)."""


def validate_tool_meta(meta: dict[str, Any] | None, *, tool_name: str = "") -> None:
    """Raise ``MetaValidationError`` unless ``meta`` carries a valid ``tier`` AND
    ``scope`` (C-TOOL). A tool with no ``_meta`` at all is rejected.

    ``tool_name`` is only used to make the error message actionable.
    """
    label = f" for tool {tool_name!r}" if tool_name else ""
    if not isinstance(meta, dict):
        raise MetaValidationError(
            f"_meta is required{label}: must declare both 'tier' and 'scope'"
        )

    tier = meta.get("tier")
    scope = meta.get("scope")
    if tier is None:
        raise MetaValidationError(f"_meta.tier is required{label} (one of {sorted(TIERS)})")
    if scope is None:
        raise MetaValidationError(f"_meta.scope is required{label} (one of {sorted(SCOPES)})")
    if tier not in TIERS:
        raise MetaValidationError(
            f"_meta.tier {tier!r} invalid{label}: must be one of {sorted(TIERS)}"
        )
    if scope not in SCOPES:
        raise MetaValidationError(
            f"_meta.scope {scope!r} invalid{label}: must be one of {sorted(SCOPES)}"
        )

    undo = meta.get("undo_hint")
    if undo is not None:
        if not isinstance(undo, dict) or "tool" not in undo:
            raise MetaValidationError(
                f"_meta.undo_hint{label} must be a dict with at least a 'tool' key"
            )

    ncf = meta.get("no_context_fill")
    if ncf is not None and not (
        isinstance(ncf, (list, tuple)) and all(isinstance(a, str) for a in ncf)
    ):
        raise MetaValidationError(
            f"_meta.no_context_fill{label} must be a list of argument names"
        )

    xa = meta.get("exclusive_args")
    if xa is not None and not (
        isinstance(xa, (list, tuple))
        and all(isinstance(g, (list, tuple)) and len(g) >= 2
                and all(isinstance(a, str) for a in g) for g in xa)
    ):
        raise MetaValidationError(
            f"_meta.exclusive_args{label} must be a list of groups, each two or more "
            "argument names"
        )

    synonyms = meta.get("synonyms")
    if synonyms is not None and not (
        isinstance(synonyms, (list, tuple))
        and all(isinstance(s, str) for s in synonyms)
    ):
        raise MetaValidationError(f"_meta.synonyms{label} must be a list of strings")


def require_meta(
    tier: str,
    scope: str,
    *,
    undo_hint: dict[str, Any] | None = None,
    synonyms: list[str] | None = None,
    async_job: bool = False,
    paid: bool = False,
    ambient_book: bool = False,
    ambient_project: bool = False,
    no_context_fill: list[str] | None = None,
    exclusive_args: list[list[str]] | None = None,
    visibility: str | None = None,
    superseded_by: str | None = None,
    tool_name: str = "",
) -> dict[str, Any]:
    """Build a validated ``_meta`` dict, ready to pass as the ``meta=`` argument
    of ``@server.tool(...)``. Raises ``MetaValidationError`` if tier/scope are
    invalid — so a misdeclared tool fails at registration time, not at call time.

    ``async_job=True`` marks a tool that STARTS a background job (queued; not done
    when the call returns) — the durable async-honesty signal a consumer (the
    workflow step-runner) reads from the catalog instead of guessing from the name.

    ``paid=True`` marks a tool whose call SPENDS REAL MONEY (Track D CD1). It is
    ORTHOGONAL to ``tier``: spend governs money, tier governs mutation. A paid READ
    (e.g. web search) stays tier ``R`` and remains callable in ``ask`` mode, but must
    clear a SPEND gate — never a write gate. Do not coerce a tool to ``A``/``W``
    merely because it costs money.

    ``no_context_fill=["book_id"]`` names arguments a consumer must NOT backfill from
    session context. A backfiller (chat-service's ``_inject_context_ids``) helpfully
    supplies a known id the model omitted, which is right for almost every tool and wrong
    for two measured shapes:

      * the argument SELECTS A CODE PATH rather than a scope —
        ``composition_motif_link_edit``'s ``book_id`` switches it from "link two motifs you
        own" to "link two motifs shared into that book". The runtime filled the omitted id,
        the tool refused, its refusal said to call again WITHOUT it, and the runtime put it
        back, so the remedy could never be followed.
      * the ambient value is the WRONG OBJECT —
        ``composition_entity_override_edit`` needs the DERIVATIVE Work's ``project_id``, and
        a book's ambient project is its CANONICAL Work. Filling it hands the model an id
        that is refused by definition, before it can look the right one up.

    Both reduce to one rule: AN ARGUMENT THE RUNTIME MAY SUPPLY MUST BE ONE THE CALLER COULD
    ALSO HAVE SUPPLIED AND MEANT. Renamed from ``mode_selecting_args`` on 2026-08-24, which
    named only the first reason.

    ``exclusive_args=[["book_id", "project_id"]]`` names argument groups of which the tool
    accepts EXACTLY ONE. A backfiller must not COMPLETE such a group: if any member is
    already present, the others stay absent.

    🔴 MEASURED 2026-09-01, and it refuted a wording fix. ``composition_list_derivatives``
    takes exactly one of ``book_id``/``project_id``, and a book turn in the studio carries
    BOTH — so ``_inject_context_ids`` filled both and the tool refused "give EXACTLY ONE".
    Live K=5: 24 of 24 calls carried both and every one was refused; store-wide, that shape
    is 0 done in 46 attempts. The turn before this was spent rewriting the refusal to say
    "call it with NO ARGUMENTS", which changed nothing, BECAUSE AN OBEDIENT MODEL GETS THE
    SAME SHAPE: with no arguments the runtime supplies both. No wording can fix a shape the
    runtime constructs after the model has spoken.

    The rule this states is the converse of ``no_context_fill``'s: that one says an argument
    may be MEANINGFUL BY ITS ABSENCE, this one says a COMBINATION may be invalid. They are
    separate because a tool can want one and not the other — ``composition_generate`` is
    exclusive over ``outline_node_id``/``chapter_id`` while wanting its ids filled normally.

        ``visibility="legacy"`` DEPRECATES the tool (CAT-4, mirrors the Go kit's
    ``WithVisibility``): it stays registered + callable, but is EXCLUDED from the
    agent's discoverable set on both federation surfaces (``tool_discovery.py`` +
    ``find-tools.ts``). Pair it with ``superseded_by=<tool>`` — the tool that REPLACES
    this one — so ``tool_list``/``tool_load`` label it and an agent migrates itself.
    Use when a tool duplicates another (e.g. a thin cross-service proxy over the
    canonical owner): deprecate, never delete.
    """
    meta: dict[str, Any] = {"tier": tier, "scope": scope}
    if undo_hint is not None:
        meta["undo_hint"] = undo_hint
    if synonyms is not None:
        meta["synonyms"] = synonyms
    if async_job:
        meta["async"] = True
    if paid:
        meta["paid"] = True
    # Studio context binding (spec 2026-07-22) — the tool resolves its book_id / project_id
    # from the envelope (X-Book-Id / X-Project-Id) when the model omits it (resolve_book_scope /
    # resolve_project_scope). The chat-service surface builder reads these to drop the id from
    # `required`; only set on a tool that ACTUALLY resolves it (migration atomicity).
    # Arguments a consumer must NOT backfill from session context. See below for the two
    # measured reasons; both reduce to the same rule — an argument the runtime may supply must
    # be one the caller could also have supplied and MEANT.
    if no_context_fill:
        meta["no_context_fill"] = list(no_context_fill)
    # Groups of which the tool accepts EXACTLY ONE. A backfiller must not complete one.
    if exclusive_args:
        meta["exclusive_args"] = [list(g) for g in exclusive_args]
    if ambient_book:
        meta["ambient_book"] = True
    if ambient_project:
        meta["ambient_project"] = True
    if visibility is not None:
        meta["visibility"] = visibility
    if superseded_by is not None:
        meta["superseded_by"] = superseded_by
    validate_tool_meta(meta, tool_name=tool_name)
    return meta
